### PR TITLE
Versioning

### PR DESCRIPTION
**Pull request by blanchma**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1,12 +1,14 @@
 var _ = require('lodash'),
     path = require('path'),
     Resources = require('./resources'),
-    Resource = require('./resource');
+    Resource = require('./resource'),
+    fs = require('fs');
+;
```
Comment (pola88) [inline lib/index.js]: remove it

File: lib/index.js [hunk 2]
```diff
@@ -50,6 +52,13 @@ var Rest = function(settings) {
   this.settings = {};
   this.routes = [];
 
+  if (settings.versioning === undefined) { settings.versioning = {} }
```
Comment (pola88) [inline lib/index.js]: you can do:

```
settings.versioning = settings.versioning || {};
```

Anyway, missing ```;```

File: lib/index.js [hunk 3]
```diff
@@ -92,65 +104,136 @@ Rest.prototype = {
...
+    var base_dir = that.settings.controllers;
+
+    _.each(versions, function(version) {
+      //Search in non searched versions plus wild card version
+      var _versions = _.clone(versions);
+      if ( _versions.indexOf("*") === -1 ) { _versions.unshift("*") }
+      _versions = _versions.splice(0, _versions.indexOf(version) + 1);
+
+      //Loop until find a folder
+      while( _versions.length > 0 ) {
```
Comment (pola88) [inline lib/index.js]: why are you using a while instead of forEach?

File: lib/index.js (hunk 1 above)
Comment (blanchma) [inline lib/index.js]: But we are using fs

File: lib/index.js (hunk 2 above)
Comment (blanchma) [inline lib/index.js]: Nice!

File: lib/index.js (hunk 1 above)
Comment (pola88) [inline lib/index.js]: I meant the ';'

File: lib/index.js (hunk 3 above)
Comment (blanchma) [inline lib/index.js]: I could use forEach but include reverse which means twice iterations. Not a particular performance issue for this case but while and pop do the same

Comment (pola88) [inline lib/index.js]: :+1:

File: lib/index.js [hunk 4]
```diff
@@ -50,6 +51,13 @@ var Rest = function(settings) {
   this.settings = {};
   this.routes = [];
 
+  settings.versioning = settings.versioning || {};
+  this.vnd_header = settings.versioning.header ? settings.versioning.header.toLowerCase() : 'accept';
+  this.vnd_grab = settings.versioning.grab || /vnd.mycompany.com\+json; version=(\d+)(,|$)/;
+  this.vnd_error = settings.versioning.error || 406;
```
Comment (pola88) [inline lib/index.js]: Why we need this?

File: lib/index.js [hunk 5]
```diff
@@ -92,64 +103,145 @@ Rest.prototype = {
...
+        }
+        _path = path.join(dir, name) + ".js";
+
+        if( fs.existsSync(_path) ) {
+          loaded[version] = require(_path);
+          break;
+        }
+      };
+      if (_.isUndefined( loaded[version] ) ) {
+        return res.status(error).end();
```
Comment (pola88) [inline lib/index.js]: :warning: ```res``` is not defined. The _load function is called when the app starts.
I think you should throw an exception or an warning.
What do you think?

Comment (blanchma) [inline lib/index.js]: You're right. Actually it was what I made until my last commit but when I
fix the configurable head I also change this. My fault.

2015-06-04 18:29 GMT-03:00 pola88 <notifications@github.com>:

File: lib/index.js [hunk 6]
```diff
@@ -92,64 +103,145 @@ Rest.prototype = {
...
+          return match[1];
 
-      if(this.subfolder) {
-        dir = path.join(dir, this.subfolder);
-      }
+        } else if (match[1] !== "" && versions[match[1]] === undefined && versions["*"] ) {
+          return "*";
 
-      this._loaded[nested_name] = require(path.join(dir, name));
+        } else if ( match && match[1] ) {
```
Comment (pola88) [inline lib/index.js]: why are you rechecking this condition?

Comment (blanchma) [inline lib/index.js]: Fixed. I change the conditionals there and this check is redundant

**Comment by pola88** [comment]
:+1:

**Pull request merged**